feat(embedded): Add setDataMask API method

gerbermichi opened the pull request:
## **User description**
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
Adds a new `setDataMask` method to the embedded dashboard API, allowing parent applications to programmatically set filter data masks in embedded dashboards.

This complements the existing `getDataMask` method and enables bidirectional data mask control - parent applications can now both read and write data masks to control embedded dashboard filters.

The implementation dispatches `updateDataMask` actions for each filter in the provided data mask, updating the Redux store and triggering filter changes in the embedded dashboard.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
N/A - API change only

### TESTING INSTRUCTIONS
1. Set up an embedded dashboard with native filters
2. From the parent application, call `embedDashboard.setDataMask({ filterId: { filterState: { value: [...] } } })`
3. Verify that the embedded dashboard filters update accordingly
4. Call `embedDashboard.getDataMask()` to confirm the data mask was set correctly

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API


___

## **CodeAnt-AI Description**
Add setDataMask API to embedded dashboard for programmatic filter updates

### What Changed
- The embedded SDK exposes setDataMask so parent applications can programmatically set dashboard filter values on an embedded dashboard.
- The embedded page accepts setDataMask calls and applies the provided data mask to dashboard filters in a single batched update, causing associated charts to refresh with the new filter values.
- Unit tests and Jest configuration were added to validate setDataMask behavior (empty, single, and multiple-filter masks) and to enable testing in a jsdom environment.

### Impact
`✅ Can set embedded dashboard filters from parent applications`
`✅ Charts refresh after external filter updates`
`✅ Easier embedding integration testing`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

gerbermichi committed:
feat(embedded): Add setDataMask API method

gerbermichi committed:
feat(embedded): Extend SDK with setDataMask support

gerbermichi committed:
feat(embedded): Trigger chart queries on setDataMask update

gerbermichi committed:
refactor(embedded): Optimize setDataMask with react-redux batch

sadpandajoe commented:
@gerbermichi mind adding tests? or do you think the existing tests already cover all the use cases?

gerbermichi committed:
test(embedded): Add unit tests for embedDashboard setDataMask

gerbermichi committed:
fix(embedded): Ignore unknown filter ids in setDataMask

Addresses review feedback on #35754. The observeDataMask payload carries
the crossFiltersChanged/nativeFiltersChanged flags alongside the mask, so
feeding it straight back into setDataMask pushed those flags through as
filter ids. Filter ids from a different dashboard had the same problem:
updateDataMask inserted them into the store, where the active-filter
derivation treats them as globally scoped and applies them to charts.

The SDK now drops non-object entries before emitting, and the embedded
API only dispatches ids the dashboard already knows about, warning about
the rest instead of applying them silently.

rusackas reviewed (approved):
Looks like this handles the batched updates plus unknown-filter-id case cleanly, with tests to match. Approving. Thanks!

gerbermichi committed:
fix(embedded): Acknowledge setDataMask and queue it until hydration

Send setDataMask over switchboard's get instead of emit, so an embedded
page that predates the method rejects the call instead of dropping it
silently.

On the iframe side, hold a mask requested before HYDRATE_DASHBOARD and
replay it once hydration lands, instead of discarding every filter id as
unknown against a store that has no filter entries yet.

gerbermichi committed:
fix(embedded): Replace lodash partition with plain filters in setDataMask

rusackas committed:
Merge branch 'master' into set-data-mask

rusackas commented on superset-embedded-sdk/jest.config.js:
```diff
@@ -0,0 +1,35 @@
+        jsx: 'react',
+        esModuleInterop: true,
+        allowSyntheticDefaultImports: true,
+      },
```
This one's moot now. The SDK dropped Jest for Vitest entirely (`vitest --run --dir src` in `package.json`, no `jest.config.js` or `ts-jest` left in the tree). CI's `embedded-sdk-test` job runs green on the current setup.

rusackas commented on superset-embedded-sdk/src/index.ts:
```diff
@@ -287,6 +288,7 @@ export async function embedDashboard({
     ourPort.get<string>('getDashboardPermalink', { anchor });
   const getActiveTabs = () => ourPort.get<string[]>('getActiveTabs');
   const getDataMask = () => ourPort.get<Record<string, any>>('getDataMask');
+  const setDataMask = (dataMask: Record<string, any>) => ourPort.emit('setDataMask', {dataMask});
```
Fixed on the SDK side. `setDataMask` in `superset-embedded-sdk/src/index.ts` now filters entries to `typeof mask === 'object' && mask !== null` before sending, so the `crossFiltersChanged`/`nativeFiltersChanged` booleans from `observeDataMask` never get forwarded. Covered by the change-trigger-flag test in both `index.test.ts` and `api.test.ts`.

rusackas commented on superset-frontend/src/embedded/api.tsx:
```diff
@@ -83,6 +86,14 @@ const getActiveTabs = () => store?.getState()?.dashboardState?.activeTabs || [];
+  batch(() => {
+    Object.entries(dataMask).forEach(([filterId, mask]) => {
+      store?.dispatch(updateDataMask(filterId, mask));
+    });
```
Fixed. `applyDataMask` in `api.tsx` now builds a `knownFilterIds` allowlist from the current `dataMask` and drops anything not in it, with a `logging.warn`, so unknown ids never reach `updateDataMask`. Covered by the 'ignores filter ids the dashboard does not know' test.

rusackas commented on superset-frontend/src/embedded/index.tsx:
```diff
@@ -264,6 +264,7 @@ window.addEventListener('message', function embeddedPageInitializer(event) {
     Switchboard.defineMethod('getActiveTabs', embeddedApi.getActiveTabs);
     Switchboard.defineMethod('getDataMask', embeddedApi.getDataMask);
     Switchboard.defineMethod('getChartStates', embeddedApi.getChartStates);
+    Switchboard.defineMethod('setDataMask', embeddedApi.setDataMask);
```
Fixed, and handled better than the suggested fix here would've been (that one just throws). `setDataMask` now queues the mask via `store.subscribe` and replays it once `dashboardInfo.id` shows up, instead of dropping it or racing hydration. There's a dedicated test for this ('queues the mask until the dashboard hydrates').

rusackas commented on superset-frontend/src/embedded/api.tsx:
```diff
@@ -83,6 +86,14 @@ const getActiveTabs = () => store?.getState()?.dashboardState?.activeTabs || [];
+  batch(() => {
+    Object.entries(dataMask).forEach(([filterId, mask]) => {
+      store?.dispatch(updateDataMask(filterId, mask));
+    });
```
This one's intentional, not a bug. `setDataMask` matches `updateDataMask`'s existing per-filter merge semantics everywhere else in the app (see `UPDATE_DATA_MASK` in the reducer), it's not meant to be a full replace. Matches the documented usage pattern too: read the full mask via `getDataMask`, mutate, pass it back. A `setDataMask({})` no-op is consistent with that.

rusackas commented on superset-embedded-sdk/src/index.ts:
```diff
@@ -355,6 +356,18 @@ export async function embedDashboard({
+          ([, mask]) => typeof mask === "object" && mask !== null,
+        ),
+      ),
+    });
```
Fixed. The SDK switched from `emit` to `ourPort.get`, specifically so a host gets a rejected promise instead of a silent no-op against an older embedded page. There's a direct test for it ('rejects when the embedded page does not support it').

rusackas commented on superset-frontend/src/embedded/api.tsx:
```diff
@@ -83,6 +88,32 @@ const getActiveTabs = () => store?.getState()?.dashboardState?.activeTabs || [];
+  const knownFilterIds = new Set(Object.keys(getDataMask()));
+  const [applicable, ignored] = partition(Object.entries(dataMask), ([id]) =>
+    knownFilterIds.has(id),
+  );
```
Same fix as the hydration-race thread on `index.tsx`, the queue-and-replay in `setDataMask` covers this too. Nothing's lost, it's held until hydration and applied then.

rusackas commented:
Went back through this before merging. All the CodeAnt threads that were still open turned out to already be fixed in code, the pre-hydration race, unknown-filter-id handling, change-trigger-flag stripping, and the emit vs get version-skew gap, so I replied and resolved them. Each has its own test on both the SDK and frontend side, plus the jest-config thread is moot now that the SDK's on Vitest. The one open design question (partial-update vs full-replace on `setDataMask`) is intentional, matches `updateDataMask`'s merge semantics everywhere else.

LGTM, merging.
